File: 124-binary-tree-maximum-path-sum/binary-tree-maximum-path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int maxSum=INT_MIN;
    int solve(TreeNode* root)
    {
        if(root==NULL)
        return 0;

        int Left=solve(root->left);
        int Right=solve(root->right);

        int phle_hi_milagaya=Left+Right+root->val;
        int ek_accha_hai=max(Left,Right)+root->val;
        int root_accha_hai=root->val;

        maxSum=max({maxSum,phle_hi_milagaya,ek_accha_hai,root_accha_hai});
        return max(ek_accha_hai,root_accha_hai);
    }

    int maxPathSum(TreeNode* root) 
    {
        solve(root);
        return maxSum;
    }
};

File: 124-binary-tree-maximum-path-sum/binary-tree-maximum-path-sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
private int maxPathSum;
    private int maxPathSumNodeToLeaf(TreeNode root){
        if(root == null) return 0;
        
        int leftSum = maxPathSumNodeToLeaf(root.left);
        int rightSum = maxPathSumNodeToLeaf(root.right);
        
        int maxChildSum = Math.max(leftSum, rightSum);
        maxPathSum = Math.max(Math.max(leftSum + rightSum + root.val, maxPathSum), Math.max(root.val,maxChildSum + root.val));
        
        return Math.max(maxChildSum + root.val, root.val);
    }
    
    public int maxPathSum(TreeNode root) {
        maxPathSum = Integer.MIN_VALUE;
        maxPathSumNodeToLeaf(root);
        return maxPathSum;
    }
}